### PR TITLE
[map] Viewport change event

### PR DESCRIPTION
**Pull request by hdamr**
Adds support for viewport change event (center, zoom, rotation, extent).
Adds IT for updating view states.
Fixes: https://github.com/vaadin/flow-components/issues/2501

### CODE REVIEW

File: vaadin-map-flow-parent/vaadin-map-flow/src/main/java/com/vaadin/flow/component/map/events/MapViewMoveEndEvent.java [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+     * Gets the view's updated rotation coordinates after map's "moveend" event.
+     *
+     * @return latest rotation in radians
+     */
+    public float getRotation() {
+        return rotation;
+    }
+
+    /**
+     * Gets the view's updated zoom level coordinates after map's "moveend"
```
Comment (sissbruecker) [inline vaadin-map-flow-parent/vaadin-map-flow/src/main/java/com/vaadin/flow/component/map/events/MapViewMoveEndEvent.java:54]:
```suggestion
     * Gets the view's updated zoom level after map's "moveend"
```

File: vaadin-map-flow-parent/vaadin-map-flow/src/main/resources/META-INF/resources/frontend/vaadin-map/mapConnector.js [hunk 2]
```diff
@@ -39,6 +39,26 @@ import { synchronize } from "./synchronization";
         mapElement.configuration.updateSize();
       },
     };
+
+    mapElement.configuration.on("moveend", (_event) => {
+      const view = mapElement.configuration.getView();
+
```
Comment (sissbruecker) [inline vaadin-map-flow-parent/vaadin-map-flow/src/main/resources/META-INF/resources/frontend/vaadin-map/mapConnector.js:45]: Nit: could remove the newline here

File: vaadin-map-flow-parent/vaadin-map-flow/src/main/java/com/vaadin/flow/component/map/configuration/ViewExtent.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
+package com.vaadin.flow.component.map.configuration;
+
+/**
+ * Extent of the current view's state in EPSG:3857 format
+ */
+public class ViewExtent {
```
Comment (sissbruecker) [inline vaadin-map-flow-parent/vaadin-map-flow/src/main/java/com/vaadin/flow/component/map/configuration/ViewExtent.java:6]:
```suggestion
/**
 * Defines an area within a map using min/max coordinates, which are by default in EPSG:3857 format
 */
public class Extent {
```
I think an extent is more of a general thing, and not only related to the currently visible area of the viewport.

File: vaadin-map-flow-parent/vaadin-map-flow/src/main/java/com/vaadin/flow/component/map/configuration/View.java [hunk 4]
```diff
@@ -116,4 +118,25 @@ public void setZoom(float zoom) {
     public String getProjection() {
         return projection;
     }
+
+    /**
+     * Gets the coordinates of the view's extent, default value is {@code 0} for
```
Comment (sissbruecker) [inline vaadin-map-flow-parent/vaadin-map-flow/src/main/java/com/vaadin/flow/component/map/configuration/View.java:123]:
```suggestion
     * Gets the extent of the view's currently visible area, default value is {@code 0} for
```

File: vaadin-map-flow-parent/vaadin-map-flow/src/main/java/com/vaadin/flow/component/map/Map.java [hunk 5]
```diff
@@ -42,6 +47,22 @@ public Map() {
         TileLayer baseLayer = new TileLayer();
         baseLayer.setSource(source);
         setBackgroundLayer(baseLayer);
+        registerEventListeners();
+    }
+
+    private void registerEventListeners() {
+        addViewMoveEndEventListener(event -> {
```
[the remark this answers is not in the document]
Comment (hdamr) [reply vaadin-map-flow-parent/vaadin-map-flow/src/main/java/com/vaadin/flow/component/map/Map.java]: Comment added.

File: vaadin-map-flow-parent/vaadin-map-flow/src/main/java/com/vaadin/flow/component/map/configuration/View.java [hunk 6]
```diff
@@ -116,4 +118,25 @@ public void setZoom(float zoom) {
...
+        return extent;
+    }
+
+    /**
+     * Sets the coordinate of current view's extent in EPSG:3857 format
+     *
+     * @param extent
+     *            the extent of the view in EPSG:3857 format
+     */
+    public void setExtent(ViewExtent extent) {
```
[the remark this answers is not in the document]
Comment (hdamr) [reply vaadin-map-flow-parent/vaadin-map-flow/src/main/java/com/vaadin/flow/component/map/configuration/View.java]: Now extent is updated in internal method (`updateInternalViewState`).

File: vaadin-map-flow-parent/vaadin-map-flow/src/main/java/com/vaadin/flow/component/map/events/MapViewMoveEndEvent.java [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+import com.vaadin.flow.component.ComponentEvent;
+import com.vaadin.flow.component.DomEvent;
+import com.vaadin.flow.component.EventData;
+import com.vaadin.flow.component.map.Map;
+import com.vaadin.flow.component.map.configuration.Coordinate;
+import com.vaadin.flow.component.map.configuration.ViewExtent;
+import elemental.json.JsonArray;
+
+/**
+ * Fired when map movement has ended.
```
[the remark this answers is not in the document]
Comment (hdamr) [reply vaadin-map-flow-parent/vaadin-map-flow/src/main/java/com/vaadin/flow/component/map/events/MapViewMoveEndEvent.java]: Fixed.

File: vaadin-map-flow-parent/vaadin-map-flow/src/main/java/com/vaadin/flow/component/map/events/MapViewMoveEndEvent.java [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+        super(source, fromClient);
+        this.rotation = (float) rotation;
+        this.zoom = (float) zoom;
+        this.center = new Coordinate(center.getNumber(0), center.getNumber(1));
+        this.viewExtent = new ViewExtent(extent.getNumber(0),
+                extent.getNumber(1), extent.getNumber(2), extent.getNumber(3));
+    }
+
+    /**
+     * Gets the view's updated rotation coordinates after map's "moveend" event.
```
[the remark this answers is not in the document]
Comment (hdamr) [reply vaadin-map-flow-parent/vaadin-map-flow/src/main/java/com/vaadin/flow/component/map/events/MapViewMoveEndEvent.java]: Fixed.

File: vaadin-map-flow-parent/vaadin-map-flow/src/main/java/com/vaadin/flow/component/map/events/MapViewMoveEndEvent.java (hunk 1 above)
Comment (hdamr) [reply vaadin-map-flow-parent/vaadin-map-flow/src/main/java/com/vaadin/flow/component/map/events/MapViewMoveEndEvent.java]: Fixed.

File: vaadin-map-flow-parent/vaadin-map-flow/src/main/resources/META-INF/resources/frontend/vaadin-map/mapConnector.js (hunk 2 above)
Comment (hdamr) [reply vaadin-map-flow-parent/vaadin-map-flow/src/main/resources/META-INF/resources/frontend/vaadin-map/mapConnector.js]: Fixed.

File: vaadin-map-flow-parent/vaadin-map-flow/src/main/java/com/vaadin/flow/component/map/configuration/ViewExtent.java (hunk 3 above)
Comment (hdamr) [reply vaadin-map-flow-parent/vaadin-map-flow/src/main/java/com/vaadin/flow/component/map/configuration/ViewExtent.java]: Refactored and made changes accordingly to method names.

File: vaadin-map-flow-parent/vaadin-map-flow/src/main/java/com/vaadin/flow/component/map/configuration/View.java (hunk 4 above)
Comment (hdamr) [reply vaadin-map-flow-parent/vaadin-map-flow/src/main/java/com/vaadin/flow/component/map/configuration/View.java]: Done.

**Review by sissbruecker — APPROVED** [review]

**Pull request merged**